### DISCUSSION TITLE
Avoid template rendering of action's parameter

### DISCUSSION THREAD

**Discussion by amitfin** [home-assistant/architecture#1169] [General]
#### Goal

An action should get a template as an input so the action can render it internally. For example, the action needs to supply variables to the render function. (Actual example can be found [here](https://github.com/amitfin/retry/?tab=readme-ov-file#backoff-parameter-optional).)

#### Challenge

It seems that currently action's data parameters are always getting rendered to strings before calling the action ([here](https://github.com/home-assistant/core/blob/0dd208a4b93f409cbda7bfdf40ae93d7611ce043/homeassistant/helpers/service.py#L414)).

#### What was tried

Changing the type of the parameter to string doesn't help because of the code [here](https://github.com/home-assistant/core/blob/0dd208a4b93f409cbda7bfdf40ae93d7611ce043/homeassistant/helpers/config_validation.py#L1465) and [here](https://github.com/home-assistant/core/blob/0dd208a4b93f409cbda7bfdf40ae93d7611ce043/homeassistant/helpers/config_validation.py#L782) which translates string parameters to templates when they contain a Jinja expression.

#### Ugly workaround

Use string with a special template format of square brackets `"[[ ... ]]"` instead of curly brackets `"{{ ... }}"`. The action changes the string to be a valid template string before rendering it.

#### Ask

Is there a way to avoid template rendering of action's parameter? Any thoughts or suggestions?

#### Proposal

Assuming there is currently no way to do so, here is a proposal:

1. Add `data_raw` dictionary for fields which shouldn't get rendered (allowing integrations to get the raw values). The core change is [here](https://github.com/amitfin/core/commit/d9af9e2b9e97137a009d3aff253604d91d9a4fc2).
2. Add `raw` option for the template selector. Using this option puts the value under the `data_raw` dictionary instead of the regular `data` dictionary. The frontend change is [here](https://github.com/amitfin/frontend/commit/eb9c503f02dd1430243d55805b3fc1bc6a3ba016).

(Both changes will be moved to PRs once the proposal gets approved.)

**Comment by farmio** [comment]
Same applies for `knx`. See https://github.com/home-assistant/core/pull/128901#issuecomment-2428270841

**Comment by amitfin** [comment]
Added a "Proposal" section to the original post (it seems that currently there is no way to do it).
